lib/videoroom/meeting.ex

```diff
@@ -0,0 +1,160 @@
+defmodule Videoroom.Meeting do
+  @moduledoc """
+  A Module responsible for handling a room.
+  """
+  use GenServer, restart: :transient
+
+  require Logger
+
+  alias Jellyfish.ServerMessage.PeerCrashed
+  alias Jellyfish.ServerMessage.PeerDisconnected
+  alias Jellyfish.ServerMessage.RoomCrashed
+
+  alias Jellyfish.Room
+  alias Videoroom.RoomRegistry
+
+  # Api
+
+  @spec start_link(Keyword.t()) :: GenServer.on_start()
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, name: registry_id(args[:name]))
+  end
+
+  @spec add_peer(binary()) :: {:ok, binary()} | {:error, binary()}
+  def add_peer(meeting_name) do
+    try do
+      GenServer.call(registry_id(meeting_name), :add_peer)
+    catch
+      :exit, {:noproc, _error} ->
+        {:error, "Failed to add peer"}
+    end
+  end
+
+  # Callbacks
+
+  @impl true
+  def init(args) do
+    args = %{
+      name: Keyword.fetch!(args, :name),
+      max_peers: Keyword.get(args, :max_peers)
+    }
+
+    Logger.metadata(room_name: args.name)
+
+    client = Jellyfish.Client.new()
+
+    with {:ok, notifier} <- Jellyfish.Notifier.start(),
+         {:ok, room} <- find_or_create_room(client, args) do
+      Process.monitor(notifier)
+
+      Logger.info("Created meeting")
+
+      {:ok, %{client: client, notifier: notifier, name: args.name, room_id: room.id}}
+    else
+      {:error, reason} ->
+        raise "Failed to create a meeting, reason: #{inspect(reason)}"
+    end
+  end
+
+  @spec find_or_create_room(Jellyfish.Client.t(), %{name: binary(), max_peers: integer()}) ::
+          {:ok, Room.t()} | {:error, atom() | binary()}
+  defp find_or_create_room(client, args) do
+    with {:ok, room_id} <- RoomRegistry.lookup(args.name),
+         {:ok, room} <- Room.get(client, room_id) do
+      {:ok, room}
+    else
+      {:error, :unregistered} ->
+        create_new_room(client, args)
+
+      error ->
+        handle_room_error(error, client, args)
+    end
+  end
+
+  defp create_new_room(client, args) do
+    case Room.create(client, max_peers: args.max_peers) do
+      {:ok, room} ->
+        RoomRegistry.insert_new(args.name, room.id)
+        {:ok, room}
+
+      error ->
+        error
+    end
+  end
+
+  defp handle_room_error({:error, reason} = error, client, args) do
+    if String.contains?(reason, "does not exist") do
```

Comment:
Using strings for error reasons isn't a good practice. IIUC this is something to change in the SDK, right?

Reply:
You're right. This is ugly and probably an anti-pattern. As soon as we implement errors as structured messages using protobuf in sdk, this will be fixed.